src/riscv_cpu.cpp

```diff
@@ -2162,75 +2171,97 @@ static uint32_t create_fld(int rd, int rs1)
+    rom[core_offset + (*data_pos)++] = val >> 32;
 }
 #endif
 
-static void create_csr64_recovery(uint32_t *rom, uint32_t *code_pos, uint32_t *data_pos, uint32_t csrn, uint64_t val)
+static void create_csr64_recovery(uint32_t *rom, uint32_t *code_pos, uint32_t *data_pos, uint32_t csrn, uint64_t val, uint32_t core_offset)
 {
     uint32_t data_off = sizeof(uint32_t) * (*data_pos - *code_pos);
 
     rom[(*code_pos)++] = create_auipc(1, data_off);
     rom[(*code_pos)++] = create_addi(1, data_off);
+    rom[(*code_pos)++] = create_add(1, 1, 3);
     rom[(*code_pos)++] = create_ld(1, 1);
     rom[(*code_pos)++] = create_csrrw(1, csrn);
 
-    rom[(*data_pos)++] = val & 0xFFFFFFFF;
-    rom[(*data_pos)++] = val >> 32;
+    rom[core_offset + (*data_pos)++] = val & 0xFFFFFFFF;
+    rom[core_offset + (*data_pos)++] = val >> 32;
 }
 
-static void create_reg_recovery(uint32_t *rom, uint32_t *code_pos, uint32_t *data_pos, int rn, uint64_t val)
+static void create_reg_recovery(uint32_t *rom, uint32_t *code_pos, uint32_t *data_pos, int rn, uint64_t val, uint32_t core_offset)
 {
-    uint32_t data_off = sizeof(uint32_t) * (*data_pos - *code_pos);
-
-    rom[(*code_pos)++] = create_auipc(rn, data_off);
-    rom[(*code_pos)++] = create_addi(rn, data_off);
-    rom[(*code_pos)++] = create_ld(rn, rn);
+    if(rn != 3) {
+        uint32_t data_off = sizeof(uint32_t) * (*data_pos - *code_pos);
+        rom[(*code_pos)++] = create_auipc(rn, data_off);
+        rom[(*code_pos)++] = create_addi(rn, data_off);
+        rom[(*code_pos)++] = create_add(rn, rn, 3);
+        rom[(*code_pos)++] = create_ld(rn, rn);
+    }
+    else {
+        rom[(*code_pos)++] = create_csrrw(1, 0x7b2);
+        rom[(*code_pos)++] = create_add(1, 3, 0);
+        uint32_t data_off = sizeof(uint32_t) * (*data_pos - *code_pos);
+        rom[(*code_pos)++] = create_auipc(rn, data_off);
+        rom[(*code_pos)++] = create_addi(rn, data_off);
+        rom[(*code_pos)++] = create_add(rn, rn, 1);
+        rom[(*code_pos)++] = create_csrrs(1, 0x7b2);
+        rom[(*code_pos)++] = create_ld(rn, rn);
+    }
 
-    rom[(*data_pos)++] = val & 0xFFFFFFFF;
-    rom[(*data_pos)++] = val >> 32;
+    rom[core_offset + (*data_pos)++] = val & 0xFFFFFFFF;
+    rom[core_offset + (*data_pos)++] = val >> 32;
 }
 
-static void create_io64_recovery(uint32_t *rom, uint32_t *code_pos, uint32_t *data_pos, uint64_t addr, uint64_t val)
+static void create_io64_recovery(uint32_t *rom, uint32_t *code_pos, uint32_t *data_pos, uint64_t addr, uint64_t val, uint32_t core_offset)
 {
     uint32_t data_off = sizeof(uint32_t) * (*data_pos - *code_pos);
 
     rom[(*code_pos)++] = create_auipc(1, data_off);
     rom[(*code_pos)++] = create_addi(1, data_off);
+    rom[(*code_pos)++] = create_add(1, 1, 3);
     rom[(*code_pos)++] = create_ld(1, 1);
 
-    rom[(*data_pos)++] = addr & 0xFFFFFFFF;
-    rom[(*data_pos)++] = addr >> 32;
+    rom[core_offset + (*data_pos)++] = addr & 0xFFFFFFFF;
```

Comment:
So the organization of the rom is:
[core 0 reg 0
 core 0 reg 1
  ...
 core 1 reg 0
 core 1 reg 1
 ...
]
?

Reply:
Yes.